File: 1062c.py
# C++17 implemention -> 1062c.cpp
# Input generator -> 1062c_gen.py

MOD = 1000000007

def main():
    buf = input()
    buflist = buf.split()
    n = int(buflist[0])
    q = int(buflist[1])
    buf = input()
    x = buf
    sum_list = [0] # sentinel / one indexing
    for i, deliciousness in enumerate(x):
        sum_list.append(int(deliciousness) + sum_list[i])
    enjoyment_list = [0]
    for i in range(n):
        enjoyment_list.append((enjoyment_list[i] * 2 + 1) % MOD)
    # buffer input to avoid TLE
    query_list = []
    for i in range(q):
        buf = input()
        buflist = buf.split()
        l = int(buflist[0]) # one indexing
        r = int(buflist[1]) # one indexing
        query_list.append((l, r))
    for i, query in enumerate(query_list):
        l = query[0]
        r = query[1]
        banhmi_count = r - l + 1
        delicious_count = sum_list[r] - sum_list[l - 1]
        non_delicious_count = banhmi_count - delicious_count
        enjoyment = 0
        # main part
        if delicious_count == 0:
            enjoyment = 0
        else:
            enjoyment += enjoyment_list[delicious_count]
            enjoyment += (enjoyment_list[banhmi_count] - enjoyment_list[delicious_count] - enjoyment_list[non_delicious_count])
            enjoyment = enjoyment % MOD
        print(enjoyment)

if __name__ == '__main__':
    main()

File: 1062c.cpp
// C++17 implemention of 1062c.py

#include <iostream>

long long int mod = 1000000007;

int main(){
  int n, q;
  std::cin >> n >> q;
  char x[100000];
  std::cin >> x;
  long long int sum_list[100001];
  sum_list[0] = 0;
  for (int i = 1; i <= n; i++){
    sum_list[i] = sum_list[i - 1] + (x[i - 1] - '0');
  }
  long long int enjoyment_list[100001];
  enjoyment_list[0] = 0;
  for (int i = 1; i <= 100000; i++){
    enjoyment_list[i] = (enjoyment_list[i - 1] * 2 + 1) % mod;
  }
  for (int i = 0; i < q; i++){
    int l, r;
    std::cin >> l >> r;
    int banhmi_count = r - l + 1;
    int delicious_count = sum_list[r] - sum_list[l - 1];
    int non_delicious_count = banhmi_count - delicious_count;
    int enjoyment = 0;
    // main part
    if (delicious_count == 0){
      enjoyment = 0;
    } else {
      //enjoyment += enjoyment_list[delicious_count];
      //enjoyment += enjoyment_list[banhmi_count] - enjoyment_list[delicious_count] - enjoyment_list[non_delicious_count];
      //enjoyment = enjoyment % mod;
      // Positive modulo: (x % mod + mod) % mod
      enjoyment = ((enjoyment_list[banhmi_count] - enjoyment_list[non_delicious_count]) % mod + mod) % mod;
    }
    std::cout << enjoyment << std::endl;
  }
  return 0;
}
